content/blog/2019-12-13-futil-fsm/index.md

```diff
@@ -0,0 +1,315 @@
+
+We provide abstract methods `new(name: &str) -> StateIndex`, `new_state() -> StateIndex`, `get_state(idx: StateIndex) -> &mut State` for `FSM` and `push_output(output: ValuedPort)`and `add_transition(transition: Edge) ` for `State` to generate actual FSM inner logic according to the graph we made in design review.
+
+
+
+###  FSM to RTL generation
+
+After generating an FSM, we need to translate the entire structure of inputs, outputs and states to synthesizable hardware using Verilog. This is done by breaking down a verilog file into distinct components.
+
+1. Module Declaration- Here we define the name of the module alongwith the inputs and outputs for it.
+2. Wire/reg definitions- These are internal signals that are used within the module.
+3. FSM- FSMs can be represented in Verilog using 3 `always` blocks - State transition, Next state logic, State outputs.
+
+To expand on how all the 3 `always` blocks are generated we discuss them below:
+
+1. State transition- This is pretty standard. It actually changes the state at a clock edge. Since this is a generic block it can be create without any inputs.
+2. Next state logic- This block has a bunch of cases for all the states. For each state in the FSM struct, based on the input transitions to it, we have `if else`  statements for next state logic.
+3. Output logic- This block contains output signals for each state represented by cases, similar to the previous block. In addition to having verilog statements for all the relevant outputs in the state, we also assign the rest of the outputs of the FSM to be zero for now. This is done to avoid inferred latches, which can occur if all outputs are not assigned in each state even though they don't change.
+
+We used `RcDoc` for formatting the Verilog files.
+
+
+
+## Hardest Parts
+
+1. FuTIL is implemented with [Rust](<https://www.rust-lang.org/>), so we spent some time to get familiar with the language.
+2. The design of FSM representation changes multiple times. Because the state should be store as pointer and then modified when we add transition and outputs to it. Rust will force the user use reference counter, which we did not realize at first. Also, even with reference counting, our implementation will result in endless loop when printed out. We therefore rely on Hashmap in the end.
+3. FuTIL *read* signals are not common in Verilog coding convention. We had two models in our minds, the Verilog valid/response model and our FuTIL valid/read model. We messed things up because of the existing of the two models and spent huge amount of time discussing which one should be the most ideal design.
```

Comment:
This definitely does seem to be the most sensitive part of the design!! I like where you ended up, but it seems like this would be worth discussing even more deeply.